### DISCUSSION TITLE
"Configuration" vs. "Build"

### DISCUSSION THREAD

**Discussion by starkos** [premake/premake-core#1647] [Ideas]
Right now, we use the word "configuration" for lots of things—too many, IMHO. But in particular, we use it to describe the build type...

```lua
workspace 'MyWorkspace'
   configurations { 'Debug', 'Release' }
```

…while internally we also have a configuration (or "config") object, which represents the collection of settings for workspace, project, or a specific build/platform pair contained by a workspace, project, or file. This configuration object contains a field to store the build configuration (eg. 'Debug') associated with the configuration instance.

So when someone says "configuration" or "config" it isn't always clear if they mean a build configuration ie. "Debug" or a collection of settings for a particular scope.

For the next major version I'm considering changing the first use of the term—a specific build—to "build".

```lua
workspace('MyWorkspace', function ()
   builds { 'Debug', 'Release' }
end)

if (cfg.build == 'Debug') then ... end
```

What's the wisdom of the crowd? Does that help clarify the meaning, or is it just change for the sake of change?

(FWIW, I'm also trying to consistently use the term "settings" for the stuff contained by a configuration or block—we currently tend to call that stuff "configuration" too. That seems less controversial though.)

**Comment by Enhex** [comment]
[Conan](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#settings) uses "build_type" so might follow along to minimize variations.

**Comment by samsinsane** [comment]
Visual Studio, MSBuild, xcodebuild and, I believe, XCode all use the term "configuration". I think it would be a mistake to deviate from this regardless of what other tools might do. We should use terminology that users know, especially when that terminology is consistent across the actions we support. Additionally, this change would be inconsistent with the `solution` to `workspace` change we made a few years ago.

**Reply by starkos** [reply]
I'm unclear about "would be inconsistent with the solution to workspace changes" but otherwise a fair point. Open to other suggestions to clear up our nomenclature though. Unless it's just a lost cause :)

**Reply by samsinsane** [reply]
Sorry, to clarify the workspace changes comment: It was my understanding that these changes were made to use the most common terminology among the actions, `solution` was really just a VS term, while `workspace` was the term used by everything else. So, it would be inconsistent to move away from "configuration", the term that everything seems to use, after moving to `workspace` because it was the most common term. For what it's worth, this was supposed to be more of a side note rather than a gotcha. 😆
